### PR TITLE
reuse previous evaluation

### PR DESCRIPTION
**Pull request by wongearl**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
reuse previous evaluation if audit level+version is the same as enforce level+version
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/pod-security-admission/admission/admission.go [hunk 1]
```diff
@@ -419,10 +419,11 @@ func (a *Admission) EvaluatePod(ctx context.Context, nsPolicy api.Policy, nsPoli
 	if klog.V(5).Enabled() {
 		klog.InfoS("PodSecurity evaluation", "policy", fmt.Sprintf("%v", nsPolicy), "op", attrs.GetOperation(), "resource", attrs.GetResource(), "namespace", attrs.GetNamespace(), "name", attrs.GetName())
 	}
-
+	levelVersions := make(map[api.LevelVersion]policy.AggregateCheckResult)
 	response := allowedResponse()
 	if enforce {
-		if result := policy.AggregateCheckResults(a.Evaluator.EvaluatePod(nsPolicy.Enforce, podMetadata, podSpec)); !result.Allowed {
+		result := policy.AggregateCheckResult{}
+		if result = policy.AggregateCheckResults(a.Evaluator.EvaluatePod(nsPolicy.Enforce, podMetadata, podSpec)); !result.Allowed {
```
Comment (liggitt) [inline staging/src/k8s.io/pod-security-admission/admission/admission.go:426]:
```suggestion
		result := policy.AggregateCheckResults(a.Evaluator.EvaluatePod(nsPolicy.Enforce, podMetadata, podSpec))
		if !result.Allowed {
```

File: staging/src/k8s.io/pod-security-admission/admission/admission.go [hunk 2]
```diff
@@ -432,27 +433,41 @@ func (a *Admission) EvaluatePod(ctx context.Context, nsPolicy api.Policy, nsPoli
...
+		levelVersions[nsPolicy.Enforce] = result
 	}
 
-	// TODO: reuse previous evaluation if audit level+version is the same as enforce level+version
-	if result := policy.AggregateCheckResults(a.Evaluator.EvaluatePod(nsPolicy.Audit, podMetadata, podSpec)); !result.Allowed {
+	// reuse previous evaluation if audit level+version is the same as enforce level+version
+	auditResult := policy.AggregateCheckResult{}
+	if tmpResult, ok := levelVersions[nsPolicy.Audit]; ok {
+		auditResult = tmpResult
+	} else {
```
Comment (liggitt) [inline staging/src/k8s.io/pod-security-admission/admission/admission.go:443]:
```suggestion
	auditResult, ok := levelVersions[nsPolicy.Audit]
	if !ok {
```

File: staging/src/k8s.io/pod-security-admission/admission/admission.go [hunk 3]
```diff
@@ -432,27 +433,41 @@ func (a *Admission) EvaluatePod(ctx context.Context, nsPolicy api.Policy, nsPoli
...
 
 	// avoid adding warnings to a request we're already going to reject with an error
 	if response.Allowed {
-		// TODO: reuse previous evaluation if warn level+version is the same as audit or enforce level+version
-		if result := policy.AggregateCheckResults(a.Evaluator.EvaluatePod(nsPolicy.Warn, podMetadata, podSpec)); !result.Allowed {
+		// reuse previous evaluation if warn level+version is the same as audit or enforce level+version
+		warnResult := policy.AggregateCheckResult{}
+		if tmpResult, ok := levelVersions[nsPolicy.Warn]; ok {
+			warnResult = tmpResult
+		} else {
```
Comment (liggitt) [inline staging/src/k8s.io/pod-security-admission/admission/admission.go:462]:
```suggestion
		warnResult, ok := levelVersions[nsPolicy.Warn]
		if !ok {
```

File: staging/src/k8s.io/pod-security-admission/admission/admission.go [hunk 4]
```diff
@@ -419,10 +419,11 @@ func (a *Admission) EvaluatePod(ctx context.Context, nsPolicy api.Policy, nsPoli
 	if klog.V(5).Enabled() {
 		klog.InfoS("PodSecurity evaluation", "policy", fmt.Sprintf("%v", nsPolicy), "op", attrs.GetOperation(), "resource", attrs.GetResource(), "namespace", attrs.GetNamespace(), "name", attrs.GetName())
 	}
-
+	levelVersions := make(map[api.LevelVersion]policy.AggregateCheckResult)
```
Comment (liggitt) [inline staging/src/k8s.io/pod-security-admission/admission/admission.go:422]: maybe rename to `cachedResults`?

**Review by liggitt — COMMENTED** [review]
a few stylistic suggestions, lgtm otherwise

could maybe use a unit test proving evaluation isn't run multiple times

File: staging/src/k8s.io/pod-security-admission/admission/admission.go (hunk 4 above)
Comment (wongearl) [reply staging/src/k8s.io/pod-security-admission/admission/admission.go]: ok

Comment (wongearl) [reply staging/src/k8s.io/pod-security-admission/admission/admission.go]: > maybe rename to `cachedResults`?

ok

**Comment by liggitt** [comment]
Looks like this needs to be formatted and squashed

**Comment by liggitt** [comment]
hack/update-gofmt.sh will do this for you

**Comment by liggitt** [comment]
rebased/squashed in https://github.com/kubernetes/kubernetes/pull/106095, kept author attribution to @wongearl

**Comment by liggitt** [comment]
/close
in favor of https://github.com/kubernetes/kubernetes/pull/106095

**Pull request closed without merging**